src/mapgen.cpp

```diff
@@ -1294,6 +1294,132 @@ void mapgen_function_json::load_place_mapings( JsonObject &jo, const std::string
     }
 }
 
+std::map<std::string, mapgen_palette> palettes;
+
+mapgen_palette mapgen_palette::load_temp( JsonObject &jo, const std::string &src )
+{
+    return load_internal( jo, src, false, true );
+}
+
+void mapgen_palette::load( JsonObject &jo, const std::string &src )
+{
+    mapgen_palette ret = load_internal( jo, src, true, false );
+    if( ret.id.empty() ) {
+        jo.throw_error( "Named palette needs an id" );
+    }
+
+    palettes[ ret.id ] = ret;
+}
+
+const mapgen_palette &mapgen_palette::get( const palette_id &id )
+{
+    const auto iter = palettes.find( id );
+    if( iter != palettes.end() ) {
+        return iter->second;
+    }
+
+    debugmsg( "Requested palette with unknown id %s", id.c_str() );
+    static mapgen_palette dummy;
+    return dummy;
+}
+
+void mapgen_palette::add( const palette_id &rh )
+{
+    add( get( rh ) );
+}
+
+void mapgen_palette::add( const mapgen_palette &rh )
+{
+    for( auto &placing : rh.format_placings ) {
+        format_placings[ placing.first ] = placing.second;
+    }
+    for( auto &placing : rh.format_terrain ) {
+        format_terrain[ placing.first ] = placing.second;
+    }
+    for( auto &placing : rh.format_furniture ) {
+        format_furniture[ placing.first ] = placing.second;
+    }
+}
+
+mapgen_palette mapgen_palette::load_internal( JsonObject &jo, const std::string &, bool require_id, bool allow_recur )
+{
+    mapgen_palette new_pal;
+    auto &format_placings = new_pal.format_placings;
+    auto &format_terrain = new_pal.format_terrain;
+    auto &format_furniture = new_pal.format_furniture;
+    if( require_id ) {
+        new_pal.id = jo.get_string( "id" );
+    }
+
+    if( jo.has_array( "palettes" ) ) {
+        if( allow_recur ) {
+            auto pals = jo.get_string_array( "palettes" );
+            for( auto &p : pals ) {
+                new_pal.add( p );
```

Comment:
I'm pretty sure that load order enforces regular static data before mapgen, as mapgen is actually processed on demand.
Meaning that by the time inline palettes are processed, all named palettes are loaded.

I intend to follow up on it to allow nested named palettes, but at the moment trying to nest a named palette will simply fail (because `allow_recur == false`).